Source/Core/Common/BitUtils.h

```diff
@@ -31,80 +34,245 @@ constexpr size_t BitSize() noexcept
+  static_assert(start + width <= BitSize<T>(), "Bitfield out of range");
+  static_assert(width > 0, "Bitfield is invalid (zero width)");
+  return ExtractBitsU(start, width, host);
 }
 
 ///
-/// Extracts a range of bits from a value.
+/// Extracts a sign-extended range of bits from a value.
 ///
-/// @param  src    The value to extract the bits from.
+/// @param  start  The beginning of the bit range. Bit 0 is the least-significant bit.
+/// @param  width  The width of the bit range in bits.
+/// @param  host   The host value to extract the bits from. Must satisfy std::integral.
 ///
-/// @tparam begin  The beginning of the bit range. This is inclusive.
-/// @tparam end    The ending of the bit range. This is inclusive.
-/// @tparam T      The type of the value.
-/// @tparam Result The returned result type. This is the unsigned analog
-///                of a signed type if a signed type is passed as T.
+/// @return The extracted bits.
+///
+template <std::integral T>
+constexpr std::make_signed_t<T> ExtractBitsS(const size_t start, const size_t width,
+                                             const T host) noexcept
+{
+  // This idiom was both undefined and implementation-defined behavior up until C++20.
+  // https://en.cppreference.com/w/cpp/language/operator_arithmetic#Bitwise_shift_operators
+  DEBUG_ASSERT(start + width <= BitSize<T>());  // Bitfield out of range
+  DEBUG_ASSERT(width > 0);                      // Bitfield is invalid (zero width)
+  const size_t rshift = BitSize<T>() - width;
+  const size_t lshift = rshift - start;
+  return static_cast<std::make_signed_t<T>>(host << lshift) >> rshift;
+}
+
+///
+/// Extracts a sign-extended range of bits from a value.
+///
+/// @tparam start  The beginning of the bit range. Bit 0 is the least-significant bit.
+/// @tparam width  The width of the bit range in bits.
+/// @param  host   The host value to extract the bits from. Must satisfy std::integral.
 ///
 /// @return The extracted bits.
 ///
-template <size_t begin, size_t end, typename T, typename Result = std::make_unsigned_t<T>>
-constexpr Result ExtractBits(const T src) noexcept
+template <size_t start, size_t width, std::integral T>
+constexpr std::make_signed_t<T> ExtractBitsS(const T host) noexcept
+{
+  static_assert(start + width <= BitSize<T>(), "Bitfield out of range");
+  static_assert(width > 0, "Bitfield is invalid (zero width)");
+  return ExtractBitsS(start, width, host);
+}
+
+///
+/// Inserts a bits into a value.
+///
+/// @param  start  The index of the bit. Bit 0 is the least-significant bit.
+/// @param  host   The host value to insert the bits into. Must satisfy std::integral.
+/// @param  val    The boolean value which is inserted into the host.
+///
+template <std::integral T>
+void InsertBit(size_t start, T& host, bool val) noexcept
+{
+  DEBUG_ASSERT(start < BitSize<T>());  // Bit out of range
+  const T mask = (T{1} << start);
+  host = (host & ~mask) | (-static_cast<T>(val) & mask);
```

Comment:
This generates a compile error on the windows builder (but only if it's instantiated, e.g. by #10848; see https://dolphin.ci/#/builders/21/builds/8286). Consider doing `(val ? T{-1} : T{0}) & mask` instead... or just `val ? mask : T{0}`, even. (I had some trouble understanding what was intended to be achieved here, too, since there's the implicit conversion from a bool to 0 or 1 in addition to the 1 to -1 to a full mask.)

Reply:
I neglected to unit-test the ExtractBit/InsertBit functions, didn't I?
I got this speed hack from https://graphics.stanford.edu/~seander/bithacks.html.  Negating the boolean value is a whopping 0.25 clock cycles faster than shifting the value into place while achieving the same result.  I made sure to static_cast it to type T before negating to avoid type size mismatches, but I should have guessed one of the compilers would throw a fit over negating unsigned types.  I will left-shift it to just get this over with.

Reply:
Another option would be `static_cast<std::make_signed_t<T>>(...)`, but I'd recommend adding a comment elaborating on it (and including one of the snippets I mentioned above as equivalent but slower, and maybe also that link you posted).

Reply:
I realized you don't even have to mask the shifted value for the InsertBit function because booleans are guaranteed to be 1 bit wide in C++ (the bithack was for C)